cstrml/staking/src/lib.rs

```diff
@@ -1528,31 +1537,51 @@ impl<T: Config> Module<T> {
     /// - O(1).
     /// - 0 DB entry.
     /// # </weight>
-    pub fn stake_limit_of(own_workloads: u128, _: u128) -> BalanceOf<T> {
-        // TODO: Stake limit calculation, this should be enable and adjust in olympus phase.
-        /*let total_issuance = TryInto::<u128>::try_into(T::Currency::total_issuance())
-            .ok()
-            .unwrap();
+    pub fn stage_one_stake_limit_of(own_workloads: u128) -> BalanceOf<T> {
+        if let Some(storage_stakes) = own_workloads.checked_mul(T::SPowerRatio::get()) {
+            storage_stakes.try_into().ok().unwrap()
+        } else {
+            (u64::max_value() as u128).try_into().ok().unwrap()
+        }
+    }
 
+    /// Calculate the stake limit by storage workloads, returns the stake limit value
+    ///
+    /// # <weight>
+    /// - Independent of the arguments. Insignificant complexity.
+    /// - O(1).
+    /// - 0 DB entry.
+    /// # </weight>
+    pub fn stage_two_stake_limit_of(own_workloads_in_kb: u128, total_workloads_in_kb: u128, total_stake_limit: u128) -> BalanceOf<T> {
         // total_workloads cannot be zero, or system go panic!
-        if total_workloads == 0 {
+        if total_workloads_in_kb == 0 {
             Zero::zero()
         } else {
-            let workloads_to_stakes = (( own_workloads.wrapping_mul(total_issuance) / total_workloads / 2) as u128)
-                .min(u64::max_value() as u128);
-
+            let workloads_to_stakes = ( own_workloads_in_kb.wrapping_mul(total_stake_limit) / total_workloads_in_kb) as u128;
             workloads_to_stakes.try_into().ok().unwrap()
-        }*/
+        }
+    }
 
-        // Now, we apply directly mapping algorithm for the early stage:
-        // 1. Maxwell 1.0: 1 terabytes -> 80,000 CRUs
-        // 2. Olympus 1.0: 1 terabytes -> 30 CRUs(tmp)
-        // ps: we treat 1 terabytes as 1_000_000_000_000 for make `mapping_ratio = 1`
-        if let Some(storage_stakes) = own_workloads.checked_mul(T::SPowerRatio::get()) {
-            storage_stakes.try_into().ok().unwrap()
-        } else {
-            (u64::max_value() as u128).try_into().ok().unwrap()
+    pub fn add_extra_ratio_according_to_effective_staking(total_issuance: BalanceOf<T>) -> Perbill {
+        let to_num =
+            |b: BalanceOf<T>| <T::CurrencyToVote as Convert<BalanceOf<T>, u64>>::convert(b);
+        // Multiple the percent by four, maximum value is 100%
+        let mul_four = |percent: Perbill| {
+            let mut init_percent = Perbill::zero();
+            for _ in 0..4 {
```

Comment:
Put this `4` and that `25` into config

Reply:
leave the 4 since it should be a function instead of a value.